sdv/datasets/demo.py

```diff
@@ -170,23 +381,183 @@ def get_available_demos(modality):
                 ``dataset_name``: The name of the dataset.
                 ``size_MB``: The unzipped folder size in MB.
                 ``num_tables``: The number of tables in the dataset.
-
-    Raises:
-        Error:
-            * If ``modality`` is not ``'single_table'``, ``'multi_table'`` or ``'sequential'``.
     """
     _validate_modalities(modality)
-    client = boto3.client('s3', config=Config(signature_version=SIGNATURE_VERSION))
+    contents = _list_objects(f'{modality}/')
     tables_info = defaultdict(list)
-    for item in client.list_objects(Bucket=BUCKET)['Contents']:
-        dataset_modality, dataset = item['Key'].split('/', 1)
-        if dataset_modality == modality.upper():
-            tables_info['dataset_name'].append(dataset.replace('.zip', ''))
-            headers = client.head_object(Bucket=BUCKET, Key=item['Key'])['Metadata']
-            size_mb = headers.get('size-mb', np.nan)
-            tables_info['size_MB'].append(round(float(size_mb), 2))
-            tables_info['num_tables'].append(headers.get('num-tables', np.nan))
-
-    df = pd.DataFrame(tables_info)
-    df['num_tables'] = pd.to_numeric(df['num_tables'])
-    return df
+    for dataset_name, yaml_key in _iter_metainfo_yaml_entries(contents, modality):
+        try:
+            raw = _get_data_from_bucket(yaml_key)
+            info = yaml.safe_load(raw) or {}
+
+            size_mb_val = info.get('dataset-size-mb')
+            try:
+                size_mb = float(size_mb_val) if size_mb_val is not None else np.nan
+            except (ValueError, TypeError):
+                LOGGER.info(
+                    f'Invalid dataset-size-mb {size_mb_val} for dataset '
+                    f'{dataset_name}; defaulting to NaN.'
+                )
+                size_mb = np.nan
+
+            num_tables_val = info.get('num-tables', np.nan)
+            if isinstance(num_tables_val, str):
+                try:
+                    num_tables_val = float(num_tables_val)
+                except (ValueError, TypeError):
+                    LOGGER.info(
+                        f'Could not cast num_tables_val {num_tables_val} to float for '
+                        f'dataset {dataset_name}; defaulting to NaN.'
+                    )
+                    num_tables_val = np.nan
+
+            try:
+                num_tables = int(num_tables_val) if not pd.isna(num_tables_val) else np.nan
+            except (ValueError, TypeError):
+                LOGGER.info(
+                    f'Invalid num-tables {num_tables_val} for '
+                    f'dataset {dataset_name} when parsing as int.'
+                )
+                num_tables = np.nan
```

Comment:
This could be broken down into smaller functions that do their own thing and in a similar fashion as the `download`